client/allocator/src/freeing_bump.rs

```diff
@@ -503,25 +509,53 @@ impl FreeingBumpHeapAllocator {
 	///
 	/// Returns the `bumper` from before the increase. Returns an `Error::AllocatorOutOfSpace` if
 	/// the operation would exhaust the heap.
-	fn bump(bumper: &mut u32, size: u32, heap_end: u32) -> Result<u32, Error> {
-		if *bumper + size > heap_end {
-			log::error!(
-				target: LOG_TARGET,
-				"running out of space with current bumper {}, mem size {}",
-				bumper,
-				heap_end
-			);
-			return Err(Error::AllocatorOutOfSpace)
+	fn bump(bumper: &mut u32, size: u32, memory: &mut impl Memory) -> Result<u32, Error> {
+		let required_size = u64::from(*bumper) + u64::from(size);
+
+		if required_size > memory.size() {
+			let required_pages =
+				u32::try_from((required_size + PAGE_SIZE as u64 - 1) / PAGE_SIZE as u64)
+					.map_err(|_| Error::Other("Number of required wasm pages is above u32"))?;
+
+			let pages = memory.pages();
+			let max_pages = memory.max_pages().unwrap_or(MAX_WASM_PAGES);
+
+			if pages == MAX_WASM_PAGES {
+				log::error!(target: LOG_TARGET, "Trying to grow wasm pages above maximum.",);
+
+				return Err(Error::AllocatorOutOfSpace)
+			} else {
+				// Let us growth by at least pages * 2, but ensure we stay in the allowed maximum
+				// number of pages.
+				let min_grow = min(pages * 2, max_pages);
+				let next_pages = max(min_grow, required_pages);
+
+				if required_pages > max_pages {
+					log::debug!(
```

Comment:
Because that just means that the runtime tries to allocate too much memory, but this isn't really any application error on the node side. We just prevent this.